### PR TITLE
Switch from `rust-toolchain` to `rust-toolchain.toml` so we can specify components

### PR DESCRIPTION
**Pull request by kkysen**
This switches all `rust-toolchain`s to `rust-toolchain.toml`s so we can specify components (and targets, etc. where needed).

The benefit of this is that `cargo` will automatically download the needed components and targets instead of erroring, which is very useful when upgrading our nightly.  CI breaks when this happens, and we can't simultaneously support multiple nightly versions (across branches and PRs) as we migrate.  It is still useful for the Docker containers to pre-install them, as that saves on re-downloading them every time, but it should prevent hard errors.  Furthermore, it is useful when developing locally as well.

This should (hopefully) make https://github.com/immunant/c2rust/pull/513 pass CI.

Previously, we would emit a `rust-toolchain` file in the transpiled crate if nightly features we needed.  That was simple with a `rust-toolchain` file, but with ` rust-toolchain.toml`, it's more complex, as we don't want to override fields in an existing `rust-toolchain.toml` that we don't have to.  As this is more complex and isn't currently needed for our existing tests and CI, I'll fix it in a future PR that merges/updates (i.e. override the `channel`, append to `components` and `targets`, etc.) the `rust-toolchain.toml` instead of wholly replacing it.

https://github.com/immunant/c2rust/blob/a91e8162edc65772c15967f0c6ccf3beef10245b/c2rust-transpile/src/build_files/mod.rs#L259-L265

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Docker image for c2rust CI and containerized development.
 # Multi-distro via optional BASE_IMAGE argument.
-# RUST_VER argument should be set to the contents of the rust-toolchain file in
+# RUST_VER argument should be set to the contents of the rust-toolchain.toml file in
```
Comment (fw-immunant) [inline docker/Dockerfile:3]: This seems wrong--`RUST_VER` should probably still be a version string, not a TOML document.

Comment (kkysen) [reply docker/Dockerfile:3]: Oh yeah, it should be "set to the contents of the channel in the rust-toolchain.toml file", right?

File: scripts/docker_build.sh [hunk 2]
```diff
@@ -27,8 +27,8 @@ build_image() {
     BASE_IMAGE=${1}
     IMAGE_TAG=$(echo ${BASE_IMAGE} | tr -s :/ - ) # replace colons and slashes with hyphens
 
-    # pull the rust version out of ../rust-toolchain to keep things synched
-    RUST_TOOLCHAIN_FILE="$SCRIPT_DIR/../rust-toolchain"
+    # pull the rust version out of ../rust-toolchain.toml to keep things synched
+    RUST_TOOLCHAIN_FILE="$SCRIPT_DIR/../rust-toolchain.toml"
     RUST_VER=$(cat $RUST_TOOLCHAIN_FILE | tr -d '\n')
```
Comment (fw-immunant) [inline scripts/docker_build.sh:32]: We now need a real parser here, sadly.

File: scripts/common.py [hunk 3]
```diff
@@ -98,8 +100,10 @@ class Config:
     CC_DB_JSON = "compile_commands.json"
 
     # Look up rust toolchain from repo root
-    with open(os.path.join(ROOT_DIR, "rust-toolchain")) as fh:
-        CUSTOM_RUST_NAME = fh.readline().strip()
+    rust_toolchain_str = Path(ROOT_DIR).joinpath("rust-toolchain.toml").read_text()
+    rust = toml.loads(rust_toolchain_str)
+    CUSTOM_RUST_NAME = rust["toolchain"]["channel"]
```
Comment (fw-immunant) [inline scripts/common.py:105]: We can probably extract this into its own `.py` somewhere to use as a utility in the shell scripts.

**Review by fw-immunant — CHANGES_REQUESTED** [review]
Issues commented inline.

**Comment by kkysen** [comment]
I thought I had fixed the parsing issues somewhere, but I guess I left that in a stash somewhere along with the toml merge changes.

File: docker/Dockerfile (hunk 1 above)
Comment (kkysen) [reply docker/Dockerfile]: Fixed in https://github.com/immunant/c2rust/pull/520/commits/48a9927ab2fdfd785d49dd60e6f956b387dfdef0.

File: scripts/provision_mac.sh [hunk 4]
```diff
@@ -34,10 +34,10 @@ type -P "pip3" >/dev/null || {
...
 pip3 install -r "$SCRIPT_DIR/requirements.txt" --user --disable-pip-version-check
 
-RUST_TOOLCHAIN_FILE="$SCRIPT_DIR/../rust-toolchain"
+RUST_TOOLCHAIN_FILE="$SCRIPT_DIR/../rust-toolchain.toml"
 export RUST_VER=$(cat $RUST_TOOLCHAIN_FILE | tr -d '\n')
 
 # Rust and dependencies
-RUST_TOOLCHAIN_FILE="$SCRIPT_DIR/../rust-toolchain"
+RUST_TOOLCHAIN_FILE="$SCRIPT_DIR/../rust-toolchain.toml"
 export RUST_VER=$(cat $RUST_TOOLCHAIN_FILE | tr -d '\n')
```
[the remark this answers is not in the document]
Comment (kkysen) [reply scripts/provision_mac.sh]: Deduplicated in https://github.com/immunant/c2rust/pull/520/commits/aabed6495a043693591fd989c6126a17185b5c89.
Fixed in https://github.com/immunant/c2rust/pull/520/commits/f8793f0ad885c207bb6be24d3f7bd50bffb7beda.

File: scripts/docker_build.sh (hunk 2 above)
Comment (kkysen) [reply scripts/docker_build.sh]: Fixed in https://github.com/immunant/c2rust/pull/520/commits/f8793f0ad885c207bb6be24d3f7bd50bffb7beda.

I'm using the refactored out `query_toml.py` now.

File: scripts/common.py (hunk 3 above)
Comment (kkysen) [reply scripts/common.py]: Fixed in https://github.com/immunant/c2rust/pull/520/commits/f8793f0ad885c207bb6be24d3f7bd50bffb7beda.

I extracted this into `query_toml.py` (a bit more generic) and then used it here.

File: scripts/common.py [hunk 5]
```diff
@@ -99,11 +101,7 @@ class Config:
 
     CC_DB_JSON = "compile_commands.json"
 
-    # Look up rust toolchain from repo root
-    rust_toolchain_str = Path(ROOT_DIR).joinpath("rust-toolchain.toml").read_text()
-    rust = toml.loads(rust_toolchain_str)
-    CUSTOM_RUST_NAME = rust["toolchain"]["channel"]
-
+    CUSTOM_RUST_NAME = query_toml(path=Path(ROOT_DIR).joinpath("rust-toolchain.toml"), query=("toolchain", "channel"))
```
[the remark this answers is not in the document]
Comment (kkysen) [reply scripts/common.py:104]: Why wouldn't tuples be iterable in Python?  They're more like immutable lists in Python than heterogenous lists, as lists can already be heterogenous.

**Comment by kkysen** [comment]
@thedataking, this does need a Docker update for the `toml` Python dependency actually.  Then this and https://github.com/immunant/c2rust/pull/513 will work.

File: scripts/common.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fw-immunant) [reply scripts/common.py:104]: The surprising thing to me is that homogeneous tuples are subtypes of `Iterable[ElemType]`, not just `Iterable[Any]` as all tuples are. I guess maybe tuples implement `Iterable[T]` for all `T` that are supertypes of all their element types... subtyping is hard.

[the remark this answers is not in the document]
Comment (kkysen) [reply scripts/common.py:104]: Homogenous tuples are of type `Tuple[T, ...]` and that's a subtype of `Iterable[T]` as they have an `.__iter__` method.  But don't try to reason too hard about Python's type-system, just run it and find out lol.

**Comment by kkysen** [comment]
@thedataking, CI is still failing on `import toml` meaning `pip install -r requirements.txt` still run and get into the pushed Docker containers.

File: scripts/common.py [hunk 6]
```diff
@@ -97,9 +101,7 @@ class Config:
 
     CC_DB_JSON = "compile_commands.json"
 
-    # Look up rust toolchain from repo root
-    with open(os.path.join(ROOT_DIR, "rust-toolchain")) as fh:
-        CUSTOM_RUST_NAME = fh.readline().strip()
+    CUSTOM_RUST_NAME = query_toml(path=Path(ROOT_DIR).joinpath("rust-toolchain.toml"), query=("toolchain", "channel"))
```
[the remark this answers is not in the document]
Comment (kkysen) [reply scripts/common.py:104]: `Path` is the idiomatic way in the minimum version of Python we're using, so I used that, but I didn't want to go back and change all the other `os.path` uses.

File: scripts/common.py [hunk 7]
```diff
@@ -10,11 +10,15 @@
 import argparse
 import platform
 import multiprocessing
+import toml
```
[the remark this answers is not in the document]
Comment (kkysen) [reply scripts/common.py:13]: Yeah, oops, I think this one is unused now that I moved things to `query_toml.py`.

[the remark this answers is not in the document]
Comment (kkysen) [reply scripts/common.py]: Fixed in https://github.com/immunant/c2rust/pull/520/commits/becec4f5b4e7e9cf841841d41c763df9f54bad1e.

**Pull request merged**